### ISSUE TITLE
[RLlib] Loaded Torchscript models have wrong output shape

### ISSUE THREAD

**Issue by jwallbridge** [ray-project/ray#25436]
### What happened + What you expected to happen

I have restored a checkpoint and saved my (non RNN) custom DQN model for inference using export_model. This generates a torchscript model (model.pt). One subtlety is that this model expects tensor inputs for (state, seq_lens) so ([ ], None) will throw an error. However, if an appropriate dummy input is provided, we can do a forward pass.

The issue is that this forward pass returns the wrong output shape, in this case 256, and not the model output shape.

### Versions / Dependencies

Ray 1.12.1

### Reproduction script

Here is a reproducible example for Cartpole. 
1. Train: train a custom DQN model.
2. Save: restore a checkpoint and export the Torchscript model.
3. Test: use the saved Torchscript model for inference. 

Train:

```python
import torch
from torch import nn
import ray
from ray import tune
from ray.rllib.models import ModelCatalog
from ray.rllib.models.torch.torch_modelv2 import TorchModelV2

class MyCustomModel(TorchModelV2, nn.Module):

    def __init__(self, obs_space, action_space, num_outputs, model_config, name, **kwargs):

        TorchModelV2.__init__(self, obs_space, action_space, num_outputs, model_config, name)
        nn.Module.__init__(self)

        self.hidden_layers = nn.Sequential(
            nn.Linear(4, 256),
            nn.Linear(256, num_outputs)
        )
        self.to_logits = nn.Linear(256, num_outputs)
        self.value_branch = nn.Linear(256, 1)
        self._output = None

    def forward(self, input_dict, state, seq_lens):
        inputs = input_dict["obs"]
        self._output = self.hidden_layers(inputs)
        logits = self.to_logits(self._output)
        return logits, state

    def value_function(self):
        value_out = self.value_branch(self._output)
        return torch.reshape(value_out, [-1])


if __name__ == "__main__":

    ray.shutdown()
    ray.init()
    
    ModelCatalog.register_custom_model('my_custom_model', MyCustomModel)

    config = {
        "env": "CartPole-v0",
        "model": {
            "custom_model": "my_custom_model",
        },
        "framework": "torch",
        "num_gpus": 0,
        "num_workers": 1,
    }
 
    stop = {"timesteps_total": 1e4}

    analysis = tune.run(
      "DQN",
      config = config,
      stop = stop,
      checkpoint_at_end = True
    )
```

Save:

```python
import ray
from ray import tune
from ray.rllib.agents import dqn
from ray.rllib.models import ModelCatalog

from cartpole_dqn_train import MyCustomModel

if __name__ == '__main__':

    ray.shutdown()
    ray.init()

    checkpoint_path = "DQN/DQN_CartPole-v0_85778_00000_0_2022-06-03_09-24-24/checkpoint_000010/checkpoint-10"
    export_path = "saved_models"
    
    ModelCatalog.register_custom_model('my_custom_model', MyCustomModel)
    
    config = {
        "env": "CartPole-v0",
        "model": {
            "custom_model": "my_custom_model",
        },
        "framework": "torch",
        "num_gpus": 0,
        "num_workers": 1,
    }
    
    agent = dqn.DQNTrainer(config = config)
    agent.restore(checkpoint_path)
    policy = agent.get_policy()
    policy.export_model(export_path)
```

Test:

```python
import os
import numpy as np
import torch

if __name__ == '__main__':

    model_path = 'saved_models/model.pt'
    
    model = torch.jit.load(model_path)
    model.eval()
    
    obs = torch.rand((1, 4))
    obs_dict = {"obs": obs}

    state = np.array([1.0])
    state = [torch.from_numpy(state)]

    seq_lens = np.ndarray((1,))
    seq_lens = torch.from_numpy(seq_lens)

    print(model(obs_dict, state, seq_lens)[0].shape) # = torch.Size([1, 256])
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by gjoliver** [comment]
I took a look at this, the model output shape is in fact 256.

in your example script, your logits layer has a shape of [256, num_outputs].
num_outputs in your case is 256, because we are using the size of the last hidden layer as num_outputs:
https://github.com/ray-project/ray/blob/master/rllib/algorithms/dqn/dqn_torch_policy.py#L169-L170

and by default, DQN use 1 hidden layer of size 256:
https://github.com/ray-project/ray/blob/master/rllib/algorithms/dqn/dqn.py#L132

You can try setting a different "fcnet_hidden" shape yourself and see the difference.
btw, the last hidden layer in your custom model looks a bit weird, it should have a shape of [256, 256], so that its output can be fed into logits.
I suggest you just hardcode your network shape based on your problem or the env.

let us know if you need anything else.

**Comment by gjoliver** [comment]
oh sorry for not being clear, what I meant was that the default DQN model always has 256 output logits.
but you can easily change by by setting a different config via the fcnet_hidden parameter.
you can also remove the fcnet_hidden config, and RLlib will automatically set num_outputs to action_space.n, as you can see from the code snippet I linked.
hope this clears some things.

**Comment by jwallbridge** [comment]
Thank you for the clarification @gjoliver 
It helped me clear up my misunderstanding.
